src/platforms/esp32/components/avm_builtins/i2c_driver.c

```diff
@@ -610,7 +628,58 @@ static NativeHandlerResult i2cdriver_consume_mailbox(Context *ctx)
     globalcontext_send_message(ctx->global, local_process_id, ret_msg);
     mailbox_remove_message(&ctx->mailbox, &ctx->heap);
 
-    return cmd == I2CCloseCmd ? NativeTerminate : NativeContinue;
+    return handler_result;
+}
+
+I2CAcquireResult i2c_driver_acquire(term i2c_port, i2c_port_t *i2c_num, GlobalContext *global)
+{
+    if (UNLIKELY(!term_is_pid(i2c_port))) {
+        ESP_LOGW(TAG, "Given term is not a I2C port driver.");
+        return I2CAcquireInvalidPeripheral;
+    }
+
+    int local_process_id = term_to_local_process_id(i2c_port);
+    Context *ctx = globalcontext_get_process_lock(global, local_process_id);
+
+    if (ctx->native_handler != i2cdriver_consume_mailbox || (ctx->platform_data == NULL)) {
+        ESP_LOGW(TAG, "Given term is not a I2C port driver.");
+        globalcontext_get_process_unlock(global, ctx);
+        return I2CAcquireInvalidPeripheral;
+    }
+
+    struct I2CData *i2c_data = ctx->platform_data;
+    i2c_data->ref_count++;
+
+    *i2c_num = i2c_data->i2c_num;
+
+    globalcontext_get_process_unlock(global, ctx);
+
+    return I2CAcquireOk;
+}
+
+void i2c_driver_release(term i2c_port, GlobalContext *global)
+{
+    if (UNLIKELY(!term_is_pid(i2c_port))) {
+        ESP_LOGW(TAG, "Given term is not a I2C port driver.");
+        return;
+    }
+
+    int local_process_id = term_to_local_process_id(i2c_port);
+    Context *ctx = globalcontext_get_process_lock(global, local_process_id);
+
+    if (ctx->native_handler != i2cdriver_consume_mailbox || (ctx->platform_data == NULL)) {
+        ESP_LOGW(TAG, "Given term is not a I2C port driver.");
+        globalcontext_get_process_unlock(global, ctx);
+        return;
+    }
+
+    struct I2CData *i2c_data = ctx->platform_data;
+    i2c_data->ref_count--;
+    NativeHandlerResult close_result = i2c_driver_maybe_close(ctx);
+    if (close_result == NativeTerminate) {
+        mailbox_send_term_signal(ctx, KillSignal, CLOSE_ATOM);
```

Comment:
You could have called scheduler_terminate or send a regular message (e.g. CmdClose).

The problem here is you kill with reason CLOSE_ATOM which is not NORMAL_ATOM and may have unexpected consequences.

Reply:
`scheduler_terminate` was my first idea, but it is not clear to me (and we didn't document that information), if it is thread safe or even if it can be used when `globalcontext_get_process_lock` has been called.